model_learning.py

```diff
@@ -0,0 +1,245 @@
+        high_sev = [int(state) for sym, state in zip(sample, state_list) if
+                    len(str(rev_smallmapping[sym.split('|')[0]])) == 3]
+        high_sev_states.update(high_sev)
+        sev_sinks.update(_sev_sinks)
+
+    print('Traces in sinks:', traces_in_sinks, 'Total traces:', total_traces, 'Percentage:',
+          100 * (traces_in_sinks / float(total_traces)))
+    print('Total medium-severity states:', len(med_sev_states))
+    print('Total high-severity states:', len(high_sev_states))
+    print('Total severe sinks:', len(sev_sinks))
+    return state_traces, med_sev_states, high_sev_states, sev_sinks
+
+
+# Step 6.3: Create state sequences (collecting sub-behaviors back into the same trace, augmented with state IDs)
+def make_state_sequences(episode_subsequences, state_traces):
+    state_sequences = dict()
+    counter = -1
+    for tid, (attack, episode_subsequence) in enumerate(episode_subsequences.items()):
+
+        if len(episode_subsequence) < 3:  # Discard subsequences of length < 3 (as in generate_traces)
+            continue
+        counter += 1
+
+        if '10.0.254' not in attack:  # TODO: move all these checks to the `_group_alerts_per_team` function
+            continue
+        if '147.75' in attack or '69.172' in attack:
+            continue
+
+        trace = [int(state) for state in state_traces[counter]]
+        max_services = [_most_frequent(epi[6]) for epi in episode_subsequence]
+
+        trace = trace[::-1]  # Reverse the trace from the S-PDFA back
+
+        # start_time, end_time, mcat, state_ID, mserv, list of unique signatures, (1st and last timestamp)
+        state_subsequence = [(epi[0], epi[1], epi[2], trace[i], max_services[i], epi[7], epi[8])
+                             for i, epi in enumerate(episode_subsequence)]
+
+        parts = attack.split('->')
+        team, attacker = parts[0].split('-')
+        victim, attack_num = parts[1].split('-')
+        attacker_victim = team + '-' + attacker + '->' + victim
+        attacker_victim_inv = team + '-' + victim + '->' + attacker
+        inv = False
+        if '10.0.254' in victim:
+            inv = True
+
+        if attacker_victim not in state_sequences.keys() and attacker_victim_inv not in state_sequences.keys():
+            if inv:
+                state_sequences[attacker_victim_inv] = []
+            else:
+                state_sequences[attacker_victim] = []
+        if inv:
+            state_sequences[attacker_victim_inv].extend(state_subsequence)
+            state_sequences[attacker_victim_inv].sort(key=lambda epi: epi[0])  # Sort in place based on starting times
+        else:
+            state_sequences[attacker_victim].extend(state_subsequence)
+            state_sequences[attacker_victim].sort(key=lambda epi: epi[0])  # Sort in place based on starting times
+
+    return state_sequences
+
+
+# Step 6.4: Group episodes (state sequences) per (team, victim) pair
+def group_episodes_per_av(state_sequences):
```

Comment:
About this function, you have mentioned that it can be moved to `make_state_subsequences`. I have tried doing that, however, it turns out not to be that trivial. The function `_get_attack_attempts` (part of the `make_attack_graphs`) uses `state_sequences` for (attacker, victim) pair:
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/4c4bf90d-e3ed-460e-a088-39821ffac182)

I think it would be easier to rather use `state_sequences` in this form (thus, per attacker-victim pair) and remove the  `group_episodes_by_av` (former `make_av_data`). The only place this (team, victim) dictionary is used is to get a set of all the victim (see below). However, this information can easily be obtained from `state_subsequences` when they are grouped per (attacker, victim) pair. So, I feel like `group_episodes_by_av` / `make_av_data` is redundant (I am also not sure why is it necessary to group (team, victim).
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/69105283-3508-44ab-9aa1-0cb93d4bf418)

Furthermore, technically state sequence is defined per (attacker, victim) pair in the paper, so shouldn't we be consistent in this case?
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5bd7090f-266f-49fe-9b62-ca733a602506)

Reply:
Indeed, if I simply change these places in `make_attack_graphs` function, the resulting graphs are the same and the need for `_group_episodes_per_av` (aka `make_av_data`) disappears. Can it be removed then?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/96152bc4-125b-4981-b898-ddc518969fd7)

Reply:
I think what I had suggested was to move the code that makes condensed_v_data  and condensed_a_data to where state sequences are being made. Since state sequences are for (a,v) pair, and the condensed_v_data and condensed_a_data are either on victim or on attacker basis.  In the current code, condensed_v_data and condensed_a_data are only relevant to extract attacker and victim IP, you are right. So you can make the suggested changes to make_attack_graphs(). However, I'd still like to keep condensed_v_data and condensed_a_data for future research, even if it is not used in the code yet.

Reply:
Yes, I have tried doing what you suggested, but then I ran into the issue when getting attack paths (as this part of code needs subsequences on (attacker, victim) basis, so calculating subsequences per attacker-victim pair is necessary).

In either case, as you have mentioned, I have updated this part of code, kept the `group_episodes_per_av` for the future, but removed the import (otherwise there is an unnecessary warning for unused import + the import line for `model_learning.py` is "too long")